docs/dev/modules.md

```diff
@@ -0,0 +1,105 @@
+# Cirq modules 
+
+Cirq has a modular architecture and is organized in a monorepo, all of the modules are in the same folder structure.
+Each module is structured as follows: 
+
+```
+cirq-<module-name>
+   <top level package 1> 
+   <top level package 2>
+   ...
+   setup.py
+   setup.cfg
+   requirements.txt
+   LICENSE
+   README.rst 
+...
+setup.py # metapackage 
+```
+
+Note that typically there is only a single top level package, but there might be exceptions. 
+
+The highest level module, `cirq` is an exception, as it is a metapackage, kind of a "parent" module, that only contains the set of default submodules as requirements. 
+This enables `pip install cirq` to have all the included submodules to be installed for our users.
+
+All submodules should depend on `cirq-core`, which is the central, core library for Cirq.    
+
+## Packaging 
+
+Each package gets published to PyPi as a separate package. To build all the wheel files locally, use
+
+```bash
+dev_tools/packaging/produce-package.sh ./dist `./dev_tools/packaging/generate-dev-version-id.sh`
+```
+ 
+Packages are versioned together, share the exact same version, and released together. 
+
+## Setting up a new module
+
+To setup a new module follow these steps: 
+
+1. Create the folder structure above, copy the files based on an existing module
+    1. LICENSE should be the same
+    2. README.rst will be the documentation that appears in PyPi
+    3. setup.py should specify an `install_requires` configuration that has `cirq-core=={module.version}` at the minimum
+2. Setup JSON serialization for each top level python package 
+
+
+### Setting up JSON serialization
+
+1. Add the `<top_level_package>/json_resolver_cache.py` file 
+    ```python
+    @functools.lru_cache(maxsize=1)  # coverage: ignore
```

Comment:
somewhat outside the scope of this PR: but why `maxsize=1?` It would technically be more performant to have no maxsize so it doesn't have to check/flush the cache, and we know there will ever only be one entry

Reply:
That's true. I think this was a copy-paste after @maffo's implementation on the original resolver cache - didn't think about it too much. Thanks for pointing it out - I guess we can leave it unbounded.